### ISSUE TITLE
Build fails when Box.V2 is referenced by a class library in .Net 4.0

### ISSUE THREAD

**Issue by mtsiakiris** [box/box-windows-sdk-v2#84]
Hi,

The following common error is raised when trying to build a class library that already has a reference to Box.V2 and during programming all works fine, until a build is performed.

The type or namespace name 'Box' could not be found (are you missing a using directive or an assembly reference?)

There are also many warnings about unresolved references by the Box.V2 like the following:

The primary reference "Box.V2" could not be resolved because it has an indirect dependency on the framework assembly "System.Net.Http, Version=1.5.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" which could not be resolved in the currently targeted framework. ".NETFramework,Version=v4.0". To resolve this problem, either remove the reference "Box.V2" or retarget your application to a framework version which contains "System.Net.Http, Version=1.5.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a". Globo.Box.Helper

Any ideas?

Thank you!

**Comment by brianytang** [comment]
This is likely due to dependent packages not being included. Are you referencing the sdk manually or through nuget? Nuget is the recommended option as it will automatically resolve dependencies

**Comment by mtsiakiris** [comment]
Thanks for the reply. The package is installed from nuget. All binaries are targeting .net 4.5. As I can see, many net libraries are custom built and are referenced explicitly form the source folders instead of gac. Inside nuget package folder there is a bit of mess in there. Other binaries are referenced from one folder, others from other folders! There maybe 2 problems.

1. When the project that need to use the SDK is targeted to framework 4.0 and choose from nuget to install, build fails because the referenced libraries need to utilize next gen c# features that are not present in net 4.0. That may be caused by incorrect nugget setup from your side.

2. When getting the code build from github, some libraries that you build locally are not included in the solution. Thus, build fails when the box.v2 build output is referenced by a net 4.0 project. On both above, when you switch to net 4.5 everything is ok.

In other words, there is no way to built the SDK in net 4.0 and use it by a net 4.0 application.

Build is performed by vs2012 professional.


One more question, does this box API supports enterprise box account functionality? I don't see anywhere the as_user parameter that needs to be passed in order for the admin to “act as” the specified user.


Thank you.


Sent from Windows Mail





From: brianytang
Sent: ‎Tuesday‎, ‎February‎ ‎10‎, ‎2015 ‎9‎:‎26‎ ‎PM
To: box/box-windows-sdk-v2
Cc: Marios Tsiakiris





This is likely due to dependent packages not being included. Are you referencing the sdk manually or through nuget? Nuget is the recommended option as it will automatically resolve dependencies

**Comment by brianytang** [comment]
For clarity, are all libraries in your solution managed by nuget? Or are some manually referenced?

1) The SDK was built as a portable class library that targets .net 4.0 and above. This can be confirmed by creating a 4.0 project and referencing the SDK through nuget. There is a dependency on a microsoft async library that will allow the async/await keywords to be used (despite them being introduced in 4.5). 

2. I'm not sure I understand there being libraries being built locally but not shared in the project. If you're referring to the sample projects -- those likely mean that you do not have the windows phone or windows 8 SDK tools installed. These are not required to build the SDK. 

There is a pull request that another community member has submitted that handles the as-user scenario. This has not yet been merged as we are still working to fully test it first.

**Comment by mtsiakiris** [comment]
Thank you very much for your time.
I reply in line on your response between: "---------------------------".

1) The SDK was built as a portable class library that targets .net 4.0 and above. This can be confirmed by creating a 4.0 project and referencing the SDK through nuget. There is a dependency on a microsoft async library that will allow the async/await keywords to be used (despite them being introduced in 4.5).

---------------------------

I understand this and that is exactly what I tried to do using VS 2012. Create a new, lets say windows application targeting .net 4.0. After that, through nuget, I installed the sdk that automatically added the driven references to my project. Add some code to use Box.V2 and build ok. So far, so good.
Then do the same on a class library application, then put some code inside and try to build. Build fails. When I change the target framework to 4.5, everything is ok. I attach the sample project named "TestBoxSDKNet40.zip" (~9,5Mb). Please explain why this is happening, is there something that I am missing?

---------------------------


    I'm not sure I understand there being libraries being built locally but not shared in the project. If you're referring to the sample projects -- those likely mean that you do not have the windows phone or windows 8 SDK tools installed. These are not required to build the SDK. 

---------------------------
No, I am not referring to the sample projects (W8 & WP), that projects doesn't load anyway for the reason you just mentioned. There seems to be some ambiguity regarding the Microsoft.Net.Http.x (2.2.20) versions and the need for an assembly redirect inside the configuration file (I have forced  System.Net.Http.Primitives to version 4.2.20.0 under .Net 4.5). The github project is referring to 1.5.0.0. In other words, the nuget package libraries are not the stable release of the github repository.
Am I right that you maybe building the code of Bcl, Bcl.Async, Net.Http and related libraries separately of the github box.v2 solution?
---------------------------


There is a pull request that another community member has submitted that handles the as-user scenario. This has not yet been merged as we are still working to fully test it first.
---------------------------
I have implemented the as-user functionality by using a custom implementation of Box.V2.Request.IRequestHandler. I suppose that it is not just simple for a complete solution. Is there any indication of the possible release date?

To understand my case, I need a class library in .Net 4.0 that will expose some specific functionality of the Box API and that library will be distributed and used by multiple applications. (It could be real pain to implement the api again and again for each project).

Thank you again.

Marios
---------------------------

**Comment by brianytang** [comment]
Apologies for the delay. Thanks for including the source -- I think that helped me identify the issue.

Turns out the compatibility with the portable class library was introduced as a patch to the .NET 4 framework. Additional info on this can be found in their blog post: http://blogs.msdn.com/b/dotnet/archive/2012/03/05/update-4-0-3-for-the-microsoft-net-framework-4.aspx

tldr; Support for portable class libraries was included in the 4.0.3 patch of the .NET framework. 

Is there any chance you could target that?

**Comment by mtsiakiris** [comment]
The .Net 4.5 installation already contains 4.0.1, 4.0.2 and 4.0.3 updates, like in my machine. Take a look on this: http://support.microsoft.com/kb/2600211?wa=wsignin1.0
at:
Because .NET 4.0.3 for .NET 4 was created 
after Visual Studio 11 Developer Preview was released, it cannot be used
 together with .NET 4.5 Developer Preview.

Although a very old article, is telling that if you have 4.5 installed, you must target 4.5 in order to build the code.
There is no way to force target 4.0.3 if you have 4.5 installed as far as I know using vs2012. If 4.5 is not installed, you can specify that version on vs2010.

One more thing is that you must force 4.0.3 installation on every deploy machine that has only 4.0. This is not an acceptable solution for me.

I think I will not use the sdk as it is possible to use it only with 4.5 and on special conditions under 4.0. I will go through a simplified implementation utilizing the box rest api directly instead.

ps.
I pulled the updates from git and I think you must clarify also the above information.
In other words, to use the possibility of building a solution in .net 4.0 is to have 4.0.3 and not 4.5 installed. Also, you must use vs2010 and not vs2012.

Thank you for your time.

Marios

Date: Thu, 19 Feb 2015 14:28:35 -0800

**Comment by Rod-Sychev** [comment]
I've got the same issue here but I've been trying to make it work on VS 2010.
I have installed .Net 4.03 patch, re-targeted my project however still getting the same errors.
Please help!

**Comment by Rod-Sychev** [comment]
Some further investigation:
I have installed Box.V2 version 1.1.1 and it compiled just fine.
Is this the way it supposed to be?
I'm guessing not...

**Comment by Rod-Sychev** [comment]
Another update:
Re-targeted to 4.0 and it compiled no problem with SDK version 1.1.1.
Tried to install the 1.1.4 version and it failed to compile.

**Comment by Rod-Sychev** [comment]
Turns out building solution with bunch of projects depending on the one with this Box.V2 and targeting 4.03 would fail with the errors mentioned in the first post.

**Comment by Rod-Sychev** [comment]
Any comments, Brian?

**Comment by brianytang** [comment]
@mtsiakiris - Yes I've gone ahead and updated the readme with the minimum version. I'm sorry to hear that does not meet your needs. 

@Rod-Sychev - I'm a little unclear on what's working and not working for you. Are you saying building it as a class library does not compile? Or using VS2010? VS2010 requires SP1 for portable class functionality (https://msdn.microsoft.com/en-us/library/vstudio/gg597391%28v=vs.100%29.aspx). 

Additionally, the difference for the version bump to 2.0 is because a change in the API required me to change a method signature which broke backwards compatibility. I don't believe v1.1.1 should compile while v2.0 does not

**Comment by Rod-Sychev** [comment]
It's easy:
- v 2.0 does not compile
- v 1.1.1 compiles but solutions fails to build if other projects are referencing the one with Box.V2 

I've got VS 2010 SP1, targeting .Net 4.03.

**Comment by brianytang** [comment]
If you create a brand new project and you reference the v2.0 SDK from nuget -- this does not compile?

**Comment by mtsiakiris** [comment]
Do you mean this: https://www.nuget.org/packages/Box.v2.SDK.Sample.Oauth2/1.0.0
There is no .Net 2.0 version of your Box.V2 as I can see in NuGet. Where can I get the one you are talking about?

Date: Tue, 3 Mar 2015 13:12:26 -0800

**Comment by brianytang** [comment]
@mtsiakiris - that is a 3rd party SDK created by a community member and may suit your needs but is not officially supported by Box.

The v2.0 I was referring to was the version of this Box SDK in nuget which @Rod-Sychev has said is not compiling for him.

**Comment by elveliz** [comment]
@brianytang I'm having the same issue. It is possible that the solution is described in this forum...  https://github.com/google/google-api-dotnet-client/issues/555

**Comment by jffcovey** [comment]
Running into this issue, but not on build.  I've installed Box.V2 version 2.12.1 and trying to make a call like BoxClient.SearchManager.SearchAsync("query") and getting an exception like 

Could not load file or assembly 'System.Net.Http.Primitives, Version=1.5.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)

Using VS 2015 and .NET 4.6.1

packages.config 

<packages>
  <package id="BAT" version="2.0.9" targetFramework="net461" />
  <package id="Box.V2" version="2.12.1" targetFramework="net461" />
  <package id="EntityFramework" version="6.1.3" targetFramework="net461" />
  <package id="Microsoft.Bcl" version="1.1.8" targetFramework="net461" />
  <package id="Microsoft.Bcl.Async" version="1.0.168" targetFramework="net461" />
  <package id="Microsoft.Bcl.Build" version="1.0.14" targetFramework="net461" />
  <package id="Microsoft.Net.Http" version="2.2.20" targetFramework="net461" />
  <package id="Newtonsoft.Json" version="6.0.2" targetFramework="net461" />
  <package id="Nito.AsyncEx" version="2.1.3" targetFramework="net461" />
  <package id="NLog" version="3.1.0.0" targetFramework="net461" />
  <package id="NLog.Config" version="3.1.0.0" targetFramework="net461" />
  <package id="NLog.Schema" version="3.1.0.0" targetFramework="net461" />
  <package id="RASD.Audit.EntityFramework" version="1.0.12" targetFramework="net461" />
  <package id="SlowCheetah" version="2.5.15" targetFramework="net461" />
</packages>

app.config

<runtime>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.Net.Http" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-2.2.29.0" newVersion="2.2.29.0" />
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="System.Net.Http.Primitives" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.2.20.0" newVersion="4.2.20.0" />
      </dependentAssembly>
    </assemblyBinding>

**Comment by coolboy** [comment]
Closing since this issue is old

Feel free to reopen it
